### PR TITLE
Add scroll event support

### PR DESCRIPTION
**Pull request by robert-ancell**

### CODE REVIEW

**Comment by robert-ancell** [comment]
It's not clear to me from the log what test is failing - any ideas?

File: shell/platform/linux/fl_view.cc [hunk 1]
```diff
@@ -241,6 +242,41 @@ static gboolean fl_view_button_release_event(GtkWidget* widget,
...
+  }
+
+  // TODO: See if this can be queried from the OS; this value is chosen
+  // arbitrarily to get something that feels reasonable.
+  const int kScrollOffsetMultiplier = 20;
+  scroll_delta_x *= kScrollOffsetMultiplier;
+  scroll_delta_y *= kScrollOffsetMultiplier;
+
+  fl_engine_send_mouse_pointer_event(
+      self->engine, self->button_state != 0 ? kMove : kHover,
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_view.cc:273]: IIRC a pointer signal should always be hover.

File: shell/platform/linux/fl_view.cc [hunk 2]
```diff
@@ -241,6 +242,41 @@ static gboolean fl_view_button_release_event(GtkWidget* widget,
...
+    scroll_delta_y = 1;
+  } else if (event->direction == GDK_SCROLL_LEFT) {
+    scroll_delta_x = -1;
+  } else if (event->direction == GDK_SCROLL_RIGHT) {
+    scroll_delta_x = 1;
+  }
+
+  // TODO: See if this can be queried from the OS; this value is chosen
+  // arbitrarily to get something that feels reasonable.
+  const int kScrollOffsetMultiplier = 20;
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_view.cc:268]: Should this be the same for smooth and non-smooth?

Comment (robert-ancell) [reply shell/platform/linux/fl_view.cc:290]: For my mouse I'm getting magnitudes of 1.0 for both smooth and non-smooth events. Existing GTK code that uses scroll events uses the same step size for both smooth and non-smooth.

The GTK scrolling code seems very weird, you get both smooth and non-smooth events (though I think drivers technically may not send smooth events). This means you get two events for every scroll, which seems wrong. And the first smooth event is (0,0), which means the first scroll is half the distance of the following ones. I've copied the same method that the GTK widgets use for scrolls, so if it's wrong it's something that probably needs to be fixed in GTK.

File: shell/platform/linux/fl_view.cc (hunk 1 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_view.cc:296]: If you hold down a mouse button and drag and do a scroll during that drag the button state should be `kMove`, right? I copied the GLFW/Windows code here and AFAICT they do the above in `SetEventPhaseFromCursorButtonState`.

**Comment by stuartmorgan** [comment]
> It's not clear to me from the log what test is failing - any ideas?

I'm not sure what that final output means; it looks like the tests didn't finish running though.

File: shell/platform/linux/fl_view.cc (hunk 1 above)
Comment (stuartmorgan) [reply shell/platform/linux/fl_view.cc:296]: I thought it was documented to always be hover, but I can't find that, and looking at the framework code it seems that the phase is ignored for signals. I may well be remembering an earlier iteration of the PR where it mattered, as I went through several versions of how to encode signals.

**Pull request merged**